supportedBackends/sfs/src/main/scala/cromwell/backend/sfs/SharedFileSystem.scala

```diff
@@ -99,6 +99,37 @@ trait SharedFileSystem extends PathFactory {
 
   def sharedFileSystemConfig: Config
 
+  lazy val cachedCopyDir: Option[Path] = None
+
+  private def localizePathViaCachedCopy(originalPath: Path, executionPath: Path, docker: Boolean): Try[Unit] = {
+    val action = Try {
+      createParentDirectory(executionPath, docker)
+      // Hash the parent. This will make sure bamfiles and their indexes stay in the same dir. This is not ideal. But should work.
+      // There should be no file collisions because two files with the same name cannot exist in the same parent dir.
+      // use .get . This strategy should not be used when there is no cachedCopyDir
+      val cachedCopySubDir: Path = cachedCopyDir.get.createChild(originalPath.toAbsolutePath.parent.hashCode.toString, asDirectory = true)
+
+      // By prepending the modtime we prevent collisions in the cache from files that have changed in between.
+      // Md5 is safer but much much slower and way too CPU intensive for big files.
+      val pathAndModTime: String = originalPath.lastModifiedTime.toEpochMilli.toString + originalPath.name
+      val cachedCopyPath: Path = cachedCopySubDir./(pathAndModTime)
+
+      // The following step cannot be performed multithreaded. Checking if the file exists and copying of the file should never
+      // happen simultaneously. Therefore this is synchronized using an object that only exists once in the JVM.
+      // This is probably not the proper 'akka' way of doing things. But I could not find the proper way of doing things
+      // after an extensive web search and read through the akka documentation.
+
+      SharedFileSystem.synchronized {
```

Comment:
While looking over this PR, we're a bit scared of this global lock... have you measured the impact to performance?

It would be nice to see a before/after measurement for workflow execution.

Reply:
Also, we are plotting a course towards multiple Cromwells running jobs simultaneously from the same DB ("Horicromtal"), and this lock mechanism will not work across JVMs.

Reply:
> While looking over this PR, we're a bit scared of this global lock... have you measured the impact to performance?

I am not too happy with this either. Performance-wise it is much better than repeated copying! But on the other hand this limits file localization to one thread. 
I would love to have this thread lock by based on the filepath, so then it only locks when it gets requests for the same file (which is good) but not one another file is copied.

I haven't figured out to do this, and I was wondering if you have any suggestions?

Reply:
> Also, we are plotting a course towards multiple Cromwells running jobs simultaneously from the same DB ("Horicromtal"), and this lock mechanism will not work across JVMs.
Thanks for letting me know.
Hmm `.lock` files are to slow. So this is going to be difficult. Maybe we can use the database?

Reply:
@aednichols I have put some thought to this and have come to the following conclusions:

1. The global lock is okay. As long as it does not include the actual copying step. The performance impact of a few instructions/clock cycles spent in a lock is minimal.
Performance-wise cached-copy is much better than repeated copying. But it is a shame that localization of `<BWA index>` has to wait for the copying of `<Reference genome>`. These files shouldn't wait on eachother.
2. `Horicromtal` is only going to be a problem when these instances share the same execution folder. But this is not a far-fetched use-case. So:
  * Communicating between cromwell processes via the database will make the code much more complicated and harder to maintain. It will also complicate the database structure.
  * The filesystem is not fast enough for locking files **within** a cromwell process, because many localizations in scatters are mostly started at almost the same time. However between processes the chance that two cromwell processes need the same file within the interval it takes to create a lock file is negligible.

So in conclusion: I am going to change the code so the locks will be based on a per originalPath bases. Between cromwellprocesses lock files are fast enough, so these will be implemented. Within cromwell processes this is not fast enough. An in-memory dictionary will be used to keep track of paths that are being copied.